Source/WebCore/platform/graphics/texmap/coordinated/CoordinatedPlatformLayerBufferVideo.cpp

```diff
@@ -0,0 +1,192 @@
+            // then directly produces RGBA data. Textures for other planes aren't used, but
+            // that's decoder's problem. We have to treat that Y texture as having RGBA data.
+            return CoordinatedPlatformLayerBufferRGB::create(*reinterpret_cast<GLuint*>(m_videoFrame.data[0]), m_size, m_flags, nullptr);
+        }
+
+        unsigned numberOfPlanes = GST_VIDEO_INFO_N_PLANES(&m_videoFrame.info);
+        std::array<GLuint, 4> planes;
+        std::array<unsigned, 4> yuvPlane;
+        std::array<unsigned, 4> yuvPlaneOffset;
+        for (unsigned i = 0; i < numberOfPlanes; ++i)
+            planes[i] = *static_cast<GLuint*>(m_videoFrame.data[i]);
+        for (unsigned i = 0; i < numberOfPlanes; ++i) {
+            yuvPlane[i] = GST_VIDEO_INFO_COMP_PLANE(&m_videoFrame.info, i);
+            yuvPlaneOffset[i] = GST_VIDEO_INFO_COMP_POFFSET(&m_videoFrame.info, i);
+        }
+
+        // Default to bt601. This is the same behaviour as GStreamer's glcolorconvert element.
+        CoordinatedPlatformLayerBufferYUV::YuvToRgbColorSpace yuvToRgbColorSpace = CoordinatedPlatformLayerBufferYUV::YuvToRgbColorSpace::BT601;
+        if (gst_video_colorimetry_matches(&GST_VIDEO_INFO_COLORIMETRY(&m_videoFrame.info), GST_VIDEO_COLORIMETRY_BT709))
+            yuvToRgbColorSpace = CoordinatedPlatformLayerBufferYUV::YuvToRgbColorSpace::BT709;
+        else if (gst_video_colorimetry_matches(&GST_VIDEO_INFO_COLORIMETRY(&m_videoFrame.info), GST_VIDEO_COLORIMETRY_BT2020))
+            yuvToRgbColorSpace = CoordinatedPlatformLayerBufferYUV::YuvToRgbColorSpace::BT2020;
+        else if (gst_video_colorimetry_matches(&GST_VIDEO_INFO_COLORIMETRY(&m_videoFrame.info), GST_VIDEO_COLORIMETRY_SMPTE240M))
+            yuvToRgbColorSpace = CoordinatedPlatformLayerBufferYUV::YuvToRgbColorSpace::SMPTE240M;
+
+        return CoordinatedPlatformLayerBufferYUV::create(numberOfPlanes, WTFMove(planes), WTFMove(yuvPlane), WTFMove(yuvPlaneOffset), yuvToRgbColorSpace, m_size, m_flags, nullptr);
+    }
+
+    return nullptr;
+}
+
+std::unique_ptr<CoordinatedPlatformLayerBuffer> CoordinatedPlatformLayerBufferVideo::createBufferFromMainMemory(GstVideoInfo* videoInfo)
+{
+    m_isMapped = gst_video_frame_map(&m_videoFrame, videoInfo, m_gstBuffer.get(), GST_MAP_READ);
+    if (!m_isMapped)
+        return nullptr;
+
+    if (GST_VIDEO_INFO_HAS_ALPHA(&m_videoFrame.info))
+        m_flags.add({ TextureMapperFlags::ShouldBlend, TextureMapperFlags::ShouldPremultiply });
+
+    auto* meta = gst_buffer_get_video_gl_texture_upload_meta(m_gstBuffer.get());
+    if (!meta || meta->n_textures != 1) // BRGx & BGRA formats use only one texture.
+        return nullptr;
+
+    OptionSet<BitmapTexture::Flags> textureFlags;
+    if (GST_VIDEO_INFO_HAS_ALPHA(&m_videoFrame.info))
+        textureFlags.add(BitmapTexture::Flags::SupportsAlpha);
+    auto texture = BitmapTexture::create(m_size, textureFlags);
```

Comment:
you're now creating a new BitmapTexture per frame, which wasn't happening before as the previous ones would be reused during the playback. What's the reason for this? Just making the code simpler? Cause this is probably going to have a performance penalty

Reply:
I think I forgot to move this to paintToTextureMapper like the other case. Good catch!

Reply:
Moved